strawberryfields/engine.py

```diff
@@ -168,11 +174,108 @@ def _run_program_locally(self, prog, **kwargs):
+            Result: results of the computation
+        """
+        if not isinstance(program, Sequence):
+            program = [program]
+
+        try:
+            prev = self.run_progs[-1] if self.run_progs else None  # previous program segment
+            for p in program:
+                if prev is None:
+                    # initialize the backend
+                    self._init_backend(p.init_num_subsystems)
+                else:
+                    # there was a previous program segment
+                    if not p.can_follow(prev):
+                        raise RuntimeError("Register mismatch: program {}, '{}'.".format(len(self.run_progs), p.name))
+
+                    # Copy the latest measured values in the RegRefs of p.
+                    # We cannot copy from prev directly because it could be used in more than one engine.
+                    for k, v in enumerate(self.samples):
+                        p.reg_refs[k].val = v
+
+                # if the program hasn't been compiled for this backend, do it now
+                if compile and p.backend != self.backend_name:
+                    p = p.compile(self.backend_name)
+                p.lock()
+
+                self._run_program(p, **kwargs)
+                self.run_progs.append(p)
+                # store the latest measurement results
+                self.samples = [p.reg_refs[k].val for k in sorted(p.reg_refs)]
+                prev = p
+        except Exception as e:
+            raise e
+        else:
+            # program execution was successful
+            pass
+
+        return Result(self.samples)
+
+
+class LocalEngine(BaseEngine):
+    """Local quantum program executor engine.
+
+    Executes :class:`.Program` instances on the chosen local backend, and makes
+    the results available via :class:`.Result`.
+
+    Args:
+        backend (str, BaseBackend): name of the backend, or a pre-constructed backend instance
+        backend_options (Dict[str, Any]): keyword arguments to be passed to the backend
+    """
+    def __init__(self, backend, *, backend_options={}):
+        super().__init__(backend, backend_options)
+
+        if isinstance(backend, str):
+            self.backend_name = backend
+            #: BaseBackend: backend for executing the quantum circuit
+            self.backend = load_backend(backend)
+        elif isinstance(backend, BaseBackend):
+            self.backend_name = backend._short_name
+            self.backend = backend
+        else:
+            raise TypeError('backend must be a string or a BaseBackend instance.')
+
+    def __str__(self):
+        return self.__class__.__name__ + '({})'.format(self.backend_name)
+
+    def reset(self, backend_options={}):
+        super().reset(backend_options)
+        self.backend_options.pop('batch_size', None)  # HACK to make tests work for now
+        self.backend.reset(**self.backend_options)
+        # TODO should backend.reset and backend.begin_circuit be combined?
```

Comment:
The difference between `begin_circuit` and `reset` is that the second allows you to specify whether the computational graph should be cleared out. Programmatically, I don't think there is huge barrier to combine them, but from user/test perspective it is cleaner to have a `reset` method (since a lot of tests need to clear out old junk from tf graph)